### PR TITLE
pattern refactor

### PR DESCRIPTION
**Pull request by doramatadora**
This introduces no breaking changes, however:

- avoids the [Promise constructor antipattern](https://stackoverflow.com/a/43050114)
- offers slight improvement to error handling
- enables strict typing
- slightly DRY-er
- removes unnecessary `n-ui` and `chai-as-promised` dependencies
- less noisy for future changes to `tsconfig.json`

No changes to test coverage.

This is in preparation for the Membership API migration to Tyk: https://github.com/Financial-Times/n-user-api-client/pull/11 (breaking change)

### CODE REVIEW

File: src/read/getUser.ts [hunk 1]
```diff
@@ -5,71 +5,42 @@ import * as R from 'ramda';
 import { readTransforms } from './transforms';
 import { ErrorWithData, errorTypes } from '../utils/error';
```
Comment (jonnywyatt) [inline src/read/getUser.ts]: logger isn't used now so can remove import

File: src/write/apis/auth-service.ts [hunk 2]
```diff
@@ -70,44 +81,42 @@ const getFetchOptions = (session: string): RequestInit => ({
...
 
-export const getAuthToken = ({ session, apiHost, apiClientId }) => {
-	return new Promise(async (resolve, reject) => {
-		try {
-			const params = {
-				response_type: 'token',
-				client_id: apiClientId,
-				scope: 'profile_max'
-			};
+const handleError = (reject, err) => {
```
Comment (jonnywyatt) [inline src/write/apis/auth-service.ts]: It doesn't look like this is used?

File: src/write/apis/user-password.ts [hunk 3]
```diff
@@ -1,46 +1,42 @@
...
+		body: JSON.stringify(body)
+	} as RequestInit;
+
+	try {
+		const response = await fetch(url, options);
+		if (response.ok) {
+			return passwordData.newPassword;
 		}
-	});
+		throw new Error(apiErrorType(response.status));
```
Comment (jonnywyatt) [inline src/write/apis/user-password.ts]: Should we standardise on ErrorWithData for thrown errors?

File: src/read/getUser.ts (hunk 1 above)
Comment (doramatadora) [reply src/read/getUser.ts]: Oops, that was an oversight. Should we remove logger altogether or use it everywhere?

**Comment by jonnywyatt** [comment]
looks great
Same question as on the other PR - are all the whitespace / formatting changes made in this PR, applied by running tslint?

File: src/utils/error.ts [hunk 4]
```diff
@@ -1,12 +1,8 @@
 export class ErrorWithData extends Error {
-	message: string;
-	data: any;
-
-	constructor(message, data?) {
+	constructor(message: string, public data?: any) {
 		super(message);
 		Object.setPrototypeOf(this, ErrorWithData.prototype);
 		(Error as any).captureStackTrace(this, this.constructor);
```
[the remark this answers is not in the document]
Comment (doramatadora) [reply src/utils/error.ts]: it's still assigned (checked by logging errors in a [test](https://github.com/Financial-Times/n-user-api-client/blob/a055b30c07f4dc1ab2b30dffb366416679d89524/test/loginUser.spec.ts#L34)) – by the addition of the keyword `public` in the constructor args. [Typescript](https://www.stevefenton.co.uk/2013/04/stop-manually-assigning-typescript-constructor-parameters/) quirk

File: src/read/getUser.ts (hunk 1 above)
Comment (doramatadora) [reply src/read/getUser.ts]: I went with "use it everywhere"

**Pull request merged**